lib/Doctrine/ORM/Cache/DefaultQueryCache.php

```diff
@@ -348,7 +348,9 @@ public function put(QueryCacheKey $key, ResultSetMapping $rsm, $result, array $h
      * @param array                             $assoc
      * @param mixed                             $assocValue
      *
-     * @return array|null
+     * @return mixed[]|null
+     *
+     * @psalm-return array{targetEntity: string, type: mixed, list?: array<array-key, array>, identifier?: array}|null
```

Comment:
AFAIK psalm, phpstan, and phpcs understand that if we use it on `@return`.

Also `array<array-key, array>` is a bit vague, isn't it?

Reply:
> Also array<array-key, array> is a bit vague, isn't it?

woops, will change that indeed

> AFAIK psalm, phpstan, and phpcs understand that if we use it on @return.

I'm doing this out of consideration for PhpStorm users (and also it's what Psalter did naturally). I'm told PhpStorm does not understand this yet, cc @goetas with who I had that debate.

Reply:
That's probably true... 😭

Reply:
My general feel on arrays matters are:
1- If array key is ignorable, then use `type[]` syntax
2- If array key is relevant, then use `array<type, type>` syntax
3- For maps of maps, repeat the process `array<type, array<type, type>>`

Anything more complex than 2 requires a comment that describes the return. Also, this hints me that it would be a good candidate for a wrapper object (that encapsulates this logic, make them immutable if needed, only allow pin-point modifications, etc).

Finally, avoid tools specific (aka. non-standardized) markers, such as phpcbf, psalm, phpstan, etc, but rather backed by official standard published publicly, such as xUnit, coverage, etc.

Reply:
> Anything more complex than 2 requires a comment that describes the return

From the original author though, I will have a hard time commenting on that after the fact. I agree that we could require that from contributors.

> Also, this hints me that it would be a good candidate for a wrapper object (that encapsulates this logic, make them immutable if needed, only allow pin-point modifications, etc).

I love how this PR makes that kind of code smell more visible

> Finally, avoid tools specific (aka. non-standardized) markers, such as phpcbf, psalm, phpstan, etc, but rather backed by official standard published publicly, such as xUnit, coverage, etc.

My hope was to remove the prefixes once Phpstorm has caught up.

Reply:
> Finally, avoid tools specific (aka. non-standardized) markers, such as phpcbf, psalm, phpstan

My two cents: psalm prefixes are actually compatible with Phpstan and Phan for the most part. Some syntax used is inspired by the unfinished PSR-5 about PHPDoc and it's quite probable that any syntax used in static analysis will make it's way into standards in the end.
Phpstorm is lagging behind because they make the mistake of waiting for standards. It slows adoption, which in turn slows the publication of standards. Adoption by big tools is exactly what we need to help everything move.